p4-16/spec/P4-16-spec.mdk

```diff
@@ -4022,15 +4019,15 @@ select (hdr.ipv4.version) {
 
 ### Masks { #sec-cubes }
 
-The infix operator `&&&` takes two arguments of type `bit<W>` or
-serializable `enum`, and creates a value of type `set<ltype>`, where
-ltype is the type of the left argument.  The right value is used as a
+The infix operator `&&&` takes two arguments of the same numeric type (section [#sec-numeric-values]),
+and creates a value of the same type. 
+The right value is used as a
 "mask", where each bit set to `0` in the mask indicates a "don't care"
```

Comment:
2022-Mar-07 LDWG meeting discussed this section of the PR for several minutes, in particular regarding whether type `int` should be allowed as operands to a `&&&` operator, and if not, then when open source p4c today accepts surface syntax like `5 &&& 7` today, there is implicit casting occurrding, probably to some `bit<W>` type, and it seems that the bit width must be coming from the context such as `select` expression or P4 table definition of key fields.

Reply:
Suggestion from Nate: Dump intermediate output after every step of open source p4c to see what it is doing in its current implementation, in terms of perhaps adding implicit casts, and what those casts are, and where did those cast types come from?  This would at least better inform what changes to the spec we may want to make in this area.

Reply:
In general the compiler treats `int` as a type variable and tries to unify with with other types. Often this will give a unique solution for the `int`, which will insert an (implicit) cast. Sometimes there will be no solution, e.g., in an expression like `2 == 3`, and then sometimes the compiler will complain (ideally if the width is necessary at runtime).

Reply:
`2 == 3` is a comparison between `int` (unbounded integer/mathematical integer). I think it should be evaluated at compile-time. The P4 standard allows compilers to reject some programs. Only accepted programs need to execute according to the manual.

Reply:
This is just an example, there are other expressions involving ints that cannot be evaluated statically. An example is calling an external function.

Reply:
The type system should guarantee that `int` values are constants as they cannot be evaluated at runtime.